### PR TITLE
Switch to use rosa hcp cluster cluster

### PR DESCRIPTION
**Pull request by xinredhat**
This PR switches to ROSA with HCP cluster. Each time the CI just bootstraps a ROSA HCP cluster, we don't need maintain a Hypershift management Server for bootstrapping cluster

### CODE REVIEW

File: ci/images/ci-runner/hack/bin/deploy-cluster.sh [hunk 1]
```diff
@@ -14,58 +14,93 @@ source "$SCRIPT_DIR/utils.sh"
...
-    printf "Control Plane pods of HostedCluster: " | indent 4
-    kubectl -n "clusters-${CLUSTER_NAME}" get pods
+# Add a function to check all of clusteroperators till AVAILABLE are true
+check_clusteroperators() {
+    operators=$(kubectl get co -o jsonpath='{range .items[*]}{.metadata.name}{" "}{end}')
 
-    # Export kubeconfig of hosted cluster
-    hypershift create kubeconfig --name "$CLUSTER_NAME" > "${CLUSTER_NAME}.kubeconfig"
+    for operator in $operators; do
+        # need to check if the operator is avialable or not in a loop
```
Comment (Roming22) [inline ci/images/ci-runner/hack/bin/deploy-cluster.sh:31]: *available

File: ci/images/ci-runner/hack/bin/deploy-cluster.sh [hunk 2]
```diff
@@ -14,58 +14,93 @@ source "$SCRIPT_DIR/utils.sh"
...
+            if [ "$available" == "True" ]; then
+                break
+            fi
+            sleep 60
+            retries=$((retries + 1))
+            echo "Retried $retries times..."
+        done
+        # if the operator is still not available after 10 times, exit 1
+        if [ "$retries" -eq "$max_retries" ]; then
+            echo "Operator $operator is not available"
```
Comment (Roming22) [inline ci/images/ci-runner/hack/bin/deploy-cluster.sh:46]: This needs to be redirected to stderr.

File: ci/images/ci-runner/hack/bin/deploy-cluster.sh [hunk 3]
```diff
@@ -14,58 +14,93 @@ source "$SCRIPT_DIR/utils.sh"
...
-    echo "kubeadmin:${pass}"
-    echo "The following is the cluster kubeconfig"
+    printf "The following is the cluster kubeconfig:\n" | indent 2
+    cat "$KUBECONFIG"
 
-    hypershift create kubeconfig --name "$CLUSTER_NAME" | tee "$WORKSPACE/kubeconfig"
+    # Check all of clusteroperators and AVAILABLE are true
+    #Workaround: Check if apiserver is ready by calling kubectl get nodes
+    if ! timeout 300s bash -c "while ! kubectl get nodes >/dev/null 2>/dev/null; do printf '.'; sleep 10; done"; then
+        echo "API server is not ready"
```
Comment (Roming22) [inline ci/images/ci-runner/hack/bin/deploy-cluster.sh:100]: This needs to be redirected to `stderr`.

**Review by Roming22 — COMMENTED** [review]
The commit tile has a typo `swith` instead of `switch`

File: ci/hack/rosa_hcp_setup.sh [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    fi
+}
+
+create_vpc() {
+    # Create a directory for the Terraform files and navigate to it
+    mkdir hypershift-tf
+    cd hypershift-tf
+
+    # Download the setup-vpc.tf file from GitHub
+    curl -s -o setup-vpc.tf https://raw.githubusercontent.com/openshift-cs/OpenShift-Troubleshooting-Templates/master/rosa-hcp-terraform/setup-vpc.tf
```
[the remark this answers is not in the document]
Comment (xinredhat) [reply ci/hack/rosa_hcp_setup.sh]: fixed

File: ci/hack/rosa_hcp_setup.sh [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+init() {
+    # Retrieve AWS Credential file from Bitwarden
+    open_bitwarden_session
+    get_aws_credentials
+    get_rosa_token
+}
+
+prechecks() {
+    if [[ -z "${PREFIX_NAME:-}" ]]; then
+      printf "PREFIX_NAME is not set\n\n"
```
[the remark this answers is not in the document]
Comment (xinredhat) [reply ci/hack/rosa_hcp_setup.sh]: fixed

File: ci/images/ci-runner/hack/bin/cleanup-expired-clusters.sh [hunk 6]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env bash
+set -o errexit
+set -o nounset
+set -o pipefail
+set -x
```
[the remark this answers is not in the document]
Comment (xinredhat) [reply ci/images/ci-runner/hack/bin/cleanup-expired-clusters.sh:5]: fixed

File: ci/images/ci-runner/hack/bin/deploy-cluster.sh (hunk 3 above)
Comment (xinredhat) [reply ci/images/ci-runner/hack/bin/deploy-cluster.sh]: fixed

File: ci/images/ci-runner/hack/bin/cleanup-expired-clusters.sh [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    for cluster in "${clusters[@]}"; do
+        if [[ "${EXCLUDE_CLUSTER[*]}" =~ $cluster ]]; then
+            continue
+        fi
+
+        is_expired=$(is_cluster_expired "$cluster")
+        
+        if [[ "$is_expired" == "true" ]]; then
+            export CLUSTER_NAME="$cluster" 
+            export REGION="$REGION"
```
[the remark this answers is not in the document]
Comment (xinredhat) [reply ci/images/ci-runner/hack/bin/cleanup-expired-clusters.sh]: fixed

File: ci/images/ci-runner/hack/bin/deploy-cluster.sh (hunk 1 above)
Comment (xinredhat) [reply ci/images/ci-runner/hack/bin/deploy-cluster.sh]: fixed

File: ci/images/ci-runner/hack/bin/deploy-cluster.sh (hunk 2 above)
Comment (xinredhat) [reply ci/images/ci-runner/hack/bin/deploy-cluster.sh]: fixed

File: developer/hack/build_image.sh [hunk 8]
```diff
@@ -140,7 +140,7 @@ build_image() {
     context="$PROJECT_DIR"
     ;;
   esac
-  $CONTAINER_ENGINE build --file "$image_dir/Dockerfile" --pull --tag "$image_name:$TAG" "$context"
+  $CONTAINER_ENGINE build --file "$image_dir/Dockerfile" --pull --tag "$image_name:$TAG" "$context"
```
[the remark this answers is not in the document]
Comment (xinredhat) [reply developer/hack/build_image.sh]: fixed

File: ci/images/ci-runner/hack/bin/cleanup-expired-clusters.sh [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+EXCLUDE_CLUSTER=(local-cluster newci4plnsvc)
+
+fetch_bitwarden_secrets() {
+    printf "Fetch secrets from bitwarden server\n" | indent 2
+    open_bitwarden_session
+    get_aws_credentials
+    get_rosa_token
+}
+
+is_cluster_expired() {
```
[the remark this answers is not in the document]
Comment (xinredhat) [reply ci/images/ci-runner/hack/bin/cleanup-expired-clusters.sh:24]: thanks for your suggestion, fixed.

**Comment by xinredhat** [comment]
Please postpone the merger as there are issues preventing the deletion of the cluster.

File: ci/images/ci-runner/hack/bin/cleanup-expired-clusters.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (Roming22) [reply ci/images/ci-runner/hack/bin/cleanup-expired-clusters.sh:5]: Does not seem to be fixed.

[the remark this answers is not in the document]
Comment (xinredhat) [reply ci/images/ci-runner/hack/bin/cleanup-expired-clusters.sh]: My bad, it was fixed this time. Thank you for the check.

**Comment by xinredhat** [comment]
the issue is fixed, this PR is ready to merge

File: ci/hack/rosa_hcp_setup.sh [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        case $1 in
+        --prefix)
+            shift
+            export PREFIX_NAME="$1"
+            ;;
+        -r | --region)
+            shift
+            export AWS_REGION="$1"
+            ;;
+        -v | --version)
```
Comment (Roming22) [inline ci/hack/rosa_hcp_setup.sh:55]: This arg does not appear in `usage()`

File: ci/hack/rosa_hcp_setup.sh [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+            ;;
+        -d | --debug)
+            set -x
+            ;;
+        -h | --help)
+            usage
+            exit 0
+            ;;
+        *)
+            echo "Unknown argument: $1"
```
Comment (Roming22) [inline ci/hack/rosa_hcp_setup.sh:67]: Redirect to stderr

File: ci/images/ci-runner/hack/bin/deploy-cluster.sh [hunk 12]
```diff
@@ -14,58 +14,93 @@ source "$SCRIPT_DIR/utils.sh"
...
     printf "Print debug info......\n" | indent 2
+    rosa describe cluster --cluster="$CLUSTER_NAME"
+}
 
-    printf "HostedCluster CR Status: " | indent 4
-    kubectl -n clusters get hostedcluster "$CLUSTER_NAME" -o yaml
-
-    printf "Control Plane pods of HostedCluster: " | indent 4
-    kubectl -n "clusters-${CLUSTER_NAME}" get pods
+# Add a function to check all of clusteroperators till AVAILABLE are true
```
Comment (Roming22) [inline ci/images/ci-runner/hack/bin/deploy-cluster.sh:26]: `# Add a function to check all of clusteroperators are AVAILABLE`

File: ci/images/ci-runner/hack/bin/deploy-cluster.sh [hunk 13]
```diff
@@ -14,58 +14,93 @@ source "$SCRIPT_DIR/utils.sh"
...
-            echo "[ERROR] Failed to create OCP cluster." >&2
+    printf "Provision hypershift cluster...\n" | indent 2
+    rosa create cluster --cluster-name "$CLUSTER_NAME" \
+        --sts --mode=auto --oidc-config-id "24132snjs8gktc5rtufqv8hggjv3ivbt" \
+        --operator-roles-prefix plnsvc-ci --region "$REGION" --version "$OCP_VERSION" \
+        --compute-machine-type m5.2xlarge \
+        --subnet-ids="subnet-001487732ebdd14f4,subnet-0718fb663f4b97f38,subnet-0fe426997da62662c" \
+        --hosted-cp -y
+
+    printf "Wait till hypershift cluster is ready...\n" | indent 2
```
Comment (Roming22) [inline ci/images/ci-runner/hack/bin/deploy-cluster.sh:68]: till -> until?

Comment (xinredhat) [reply ci/images/ci-runner/hack/bin/deploy-cluster.sh]: Fixed. I changed it to "Track the progress of the cluster creation...\n".

File: ci/images/ci-runner/hack/bin/deploy-cluster.sh (hunk 12 above)
Comment (xinredhat) [reply ci/images/ci-runner/hack/bin/deploy-cluster.sh]: fixed

File: ci/hack/rosa_hcp_setup.sh (hunk 11 above)
Comment (xinredhat) [reply ci/hack/rosa_hcp_setup.sh]: fixed

File: ci/hack/rosa_hcp_setup.sh (hunk 10 above)
Comment (xinredhat) [reply ci/hack/rosa_hcp_setup.sh:57]: fixed

**Review by Roming22 — APPROVED** [review]

**Pull request merged**